File: 56-merge-intervals/merge-intervals.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        intervals.sort(key = lambda i : i[0]) # sort by start
        output = [intervals[0]] # save into output

        for start,end in intervals[1:]:
            lastEnd = output[-1][1] # end of last element

            if start <= lastEnd: # if current start less than prev element's end then it overlaps
                output[-1][1] =  max(lastEnd, end) #update the lastEnd to the highest no
            else:
                output.append([start,end]) # if they don't overlap then simply append to output
        return output

File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        // Sort by starting value
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        List<int[]> output = new ArrayList<>();
        output.add(intervals[0]);

        // Check remaining intervals
        for (int i = 1; i < intervals.length; i++) {

            int start = intervals[i][0];
            int end = intervals[i][1];

            // Last interval in output
            int[] last = output.get(output.size() - 1);

            // Overlap
            if (start <= last[1]) {

                // Extend the end
                last[1] = Math.max(last[1], end);

            }
            else {

                // No overlap
                output.add(new int[]{start, end});
            }
        }

        return output.toArray(new int[output.size()][]);
    }
}